test/Microsoft.TestPlatform.Client.UnitTests/Discovery/DiscoveryRequestTests.cs

```diff
@@ -278,5 +222,70 @@ public void HandleRawMessageShouldAddVSTestDataPointsIfTelemetryOptedIn()
+            this.mockDataSerializer.Setup(x => x.DeserializeMessage(It.IsAny<string>()))
+                .Returns(new Message() { MessageType = MessageType.DiscoveryComplete });
+            this.mockDataSerializer.Setup(x => x.DeserializePayload<DiscoveryCompletePayload>(It.IsAny<Message>()))
+                .Returns(new DiscoveryCompletePayload()
+                {
+                    TotalTests = 1,
+                    IsAborted = false,
+                    LastDiscoveredTests = Enumerable.Empty<TestCase>()
+                });
+
+            this.discoveryRequest.HandleRawMessage(string.Empty);
+
+            this.loggerManager.Verify(lm => lm.HandleDiscoveryComplete(It.IsAny<DiscoveryCompleteEventArgs>()), Times.Once);
+        }
+
+        [TestMethod]
+        public void HandleDiscoveryCompleteShouldInvokeHandleDiscoveryCompleteOfLoggerManager()
+        {
+            var discoveryCompleteEventArgs = new DiscoveryCompleteEventArgs(1, false);
+            var eventsHandler = this.discoveryRequest as ITestDiscoveryEventsHandler2;
+            eventsHandler.HandleDiscoveryComplete(discoveryCompleteEventArgs, Enumerable.Empty<TestCase>());
+
+            loggerManager.Verify(lm => lm.HandleDiscoveryComplete(discoveryCompleteEventArgs), Times.Once);
+        }
+
+        [TestMethod]
+        public void HandleDiscoveryCompleteShouldNotInvokeHandleDiscoveredTestsIfLastChunkNotPresent()
+        {
+            var discoveryCompleteEventArgs = new DiscoveryCompleteEventArgs(1, false);
+            var eventsHandler = this.discoveryRequest as ITestDiscoveryEventsHandler2;
+            eventsHandler.HandleDiscoveryComplete(discoveryCompleteEventArgs, Enumerable.Empty<TestCase>());
+
+            loggerManager.Verify(lm => lm.HandleDiscoveredTests(It.IsAny<DiscoveredTestsEventArgs>()), Times.Never);
+        }
+
+        [TestMethod]
+        public void HandleDiscoveryCompleteShouldInvokeHandleDiscoveredTestsIfLastChunkPresent()
+        {
+            var activeTestCases = new List<ObjectModel.TestCase>
+            {
+                new ObjectModel.TestCase(
+                    "A.C.M2",
+                    new Uri("executor://dummy"),
+                    "A")
+            };
+
+            var discoveryCompleteEventArgs = new DiscoveryCompleteEventArgs(1, false);
+            var eventsHandler = this.discoveryRequest as ITestDiscoveryEventsHandler2;
+            eventsHandler.HandleDiscoveryComplete(discoveryCompleteEventArgs, activeTestCases);
+
+            loggerManager.Verify(lm => lm.HandleDiscoveredTests(It.IsAny<DiscoveredTestsEventArgs>()), Times.Once);
+        }
+
+        [TestMethod]
+        public void HandleDiscoveredTestsShouldInvokeHandleDiscoveredTestsOfLoggerManager()
+        {
+            discoveryRequest.HandleDiscoveredTests(null);
+
+            loggerManager.Verify(lm => lm.HandleDiscoveredTests(It.IsAny<DiscoveredTestsEventArgs>()), Times.Once);
+        }
     }
 }
```

Comment:
Need test to validate that HandleDiscoveryComplete of Logger Manager is invoked only once,if the message comes from both HandleRawMessage, & HandleDiscoveryComplete

Reply:
UT are there for this scenario.